### ISSUE TITLE
Rewards actual_end index in PPOv2 trainer

### ISSUE THREAD

**Issue by zhuyuzy** [huggingface/trl#1893]
In PPOv2 trainer.train(), # 4. compute rewards, when computing the rewards index, the sequence_lengths_p1 is used. 

`actual_end = torch.where(sequence_lengths_p1 < rewards.size(1), sequence_lengths_p1, sequence_lengths)`

It means we assign the reward one step after the stop_token_id, as far as I know, the reward is usually assigned to the final token which means the exact position where stop_token_id is. Did I miss or misunderstand something?

**Comment by qgallouedec** [comment]
Thanks for raising this question.

Actually, the name `sequence_lengths` is misleading, since it's actually the sequence length - 1:

https://github.com/huggingface/trl/blob/cbcaa46cd3c02c0e7f724b764c5848ae73796de7/trl/trainer/ppov2_trainer.py#L351

Consequently, if `stop_token_id == 50277` and `postprocessed_response == [[187,   187,     3, 50277,    13,   642]]`, `sequence_lengths` (and `actual_end`) will get the value `2` (and not `3`).

**Comment by zhuyuzy** [comment]
> Thanks for raising this question.
> 
> Actually, the name `sequence_lengths` is misleading, since it's actually the sequence length - 1:
> 
> https://github.com/huggingface/trl/blob/cbcaa46cd3c02c0e7f724b764c5848ae73796de7/trl/trainer/ppov2_trainer.py#L351
> 
> Consequently, if `stop_token_id == 50277` and `postprocessed_response == [[187, 187, 3, 50277, 13, 642]]`, `sequence_lengths` (and `actual_end`) will get the value `2` (and not `3`).

Thank you for your reply.

However, it's not the same in my experiment. `first_true_indices(postprocessed_response == tokenizer.pad_token_id)` returns the first index where the `pad_token_id` is, in your case 4, then minus 1 we have the `sequence_length`, in your case 3 (not 2). And finally we have `actual_end = sequence_lengths_p1` if `stop_token_id` exists, which is `sequence_length` plus 1 (in your case 4).

**Comment by zhuyuzy** [comment]
> > Thanks for raising this question.
> > Actually, the name `sequence_lengths` is misleading, since it's actually the sequence length - 1:
> > https://github.com/huggingface/trl/blob/cbcaa46cd3c02c0e7f724b764c5848ae73796de7/trl/trainer/ppov2_trainer.py#L351
> > 
> > Consequently, if `stop_token_id == 50277` and `postprocessed_response == [[187, 187, 3, 50277, 13, 642]]`, `sequence_lengths` (and `actual_end`) will get the value `2` (and not `3`).
> 
> Thank you for your reply.
> 
> However, it's not the same in my experiment. `first_true_indices(postprocessed_response == tokenizer.pad_token_id)` returns the first index where the `pad_token_id` is, in your case 4, then minus 1 we have the `sequence_length`, in your case 3 (not 2). And finally we have `actual_end = sequence_lengths_p1` if `stop_token_id` exists, which is `sequence_length` plus 1 (in your case 4).

@qgallouedec Hello, could you help clarify this? BTW, I’ve run the experiment with setting

`actual_end = sequence_lengths`

and found out the result is almost identical with

`actual_end = torch.where(sequence_lengths_p1 < rewards.size(1), sequence_lengths_p1, sequence_lengths)`

However, I’m still puzzled about why you chose to approach it this way. 

Really appreciate it if you could provide an explanation. Thank you!

**Comment by qgallouedec** [comment]
> Consequently, if `stop_token_id == 50277` and `postprocessed_response == [[187, 187, 3, 50277, 13, 642]]`, `sequence_lengths` (and `actual_end`) will get the value `2` (and not `3`).

> However, it's not the same in my experiment. `first_true_indices(postprocessed_response == tokenizer.pad_token_id)` returns the first index where the `pad_token_id` is, in your case 4

why 4?

**Comment by GJ98** [comment]
I'm also interested in this issue!

**Comment by albertvillanova** [comment]
Hey @zhuyuzy (and @GJ98), sorry for the very slow reply — we're catching up on the backlog. 

You're both reading the code correctly: `sequence_length` is the index of the stop token, and the reward-model score is added one position later, at `sequence_lengths_p1` (the first pad slot), not exactly on the stop token. This is intentional. The value tensor deliberately keeps that one extra position valid — see `padding_mask_p1 = response_idxs > sequence_lengths_p1` at https://github.com/huggingface/trl/blob/8db673e40d81a2851b5f57931ce81c68a3849aa3/trl/experimental/ppo/ppo_trainer.py#L767-L769 — and GAE bootstraps every step with `values[t + 1]` (https://github.com/huggingface/trl/blob/8db673e40d81a2851b5f57931ce81c68a3849aa3/trl/experimental/ppo/ppo_trainer.py#L790-L793), so the terminal score lands on exactly the value slot the model is trained to predict it at. Putting it on the EOS token instead (`actual_end = sequence_lengths`) is a legitimate alternative and, as you found empirically, gives near-identical results, because with gamma=lambda=1 GAE just propagates the reward back one extra step. The placement here mirrors the reference RLHF-PPO implementation this trainer was ported from. FYI PPO now lives under `trl.experimental.ppo`. 

Closing as working-as-intended, but happy to reopen if something still looks off.
